bot/common/tasks/tasks.py

```diff
@@ -158,19 +167,106 @@ class Days:
+# TODO: future refactor; abstract bot tasks class to accept a
+# cadence, action, and loop settings (enable/disable & wakeup time)
+# TODO: loop settings into its own class
+class ReportingTask(commands.Cog):
+    REPORT_LAST_SENT_DATETIME_CACHE_KEY = "contribution_report_last_sent"
+
+    def __init__(self, bot: discord.Bot, cache: Cache, cadence: Weekly, loop_settings):
+        self.bot = bot
+        self.cache: Cache = cache
+        self.cadence: Weekly = cadence
+        self.min_time_between_loop_seconds: int = 60
+        self.init_loop(loop_settings)
+
+    def init_loop(self, loop_settings):
+        # TODO: checking on loop settings to be encapsulated in class/fn
+        if not loop_settings["enable"]:
+            logger.info("contribution report task disabled, skipping...")
+            return
+
+        self.min_time_between_loop_seconds = loop_settings[
+            "min_time_between_loop_seconds"
+        ]
+        m = loop_settings["task_wakeup_period_minutes"]
+        self.contribution_report: tasks.Loop = tasks.loop(minutes=m)(
+            self.contribution_report
+        )
+        self.contribution_report.before_loop(self.wait_until_ready)
+        self.contribution_report.start()
+        self.contribution_report.error(self.handle_error)
 
-    def inner(f):
-        task_collection.add_task(f, name, cadence)
+    async def contribution_report(self):
+        now: datetime = datetime.now()
+        logger.info(f"contribution_report awake at {now}")
+
+        td_until_run: timedelta = self.cadence.get_timedelta_until_run(now)
+        s_until_run = td_until_run.total_seconds()
+
+        if s_until_run > 0:
+            if td_until_run.days > 0:
+                logger.info("contribution_report sleeping...")
+                return
+            logger.info(f"contribution_report waiting {s_until_run}s until run")
+            await asyncio.sleep(s_until_run)
+
+        # check cache
+        last_sent = await self.cache.get(self.REPORT_LAST_SENT_DATETIME_CACHE_KEY)
+        if last_sent is not None:
+            last_sent = datetime.strptime(last_sent.decode("utf-8"), DATETIME_CACHE_FMT)
+
+        td_last_sent: timedelta = None if last_sent is None else now - last_sent
+
+        # if there's a cache entry from the same day, skip the task
+        # TODO: extract the 24h last run as a setting
+        if (
+            td_last_sent is not None
+            and td_last_sent.total_seconds() < self.min_time_between_loop_seconds
+        ):
+            logger.warn(
+                f"last sent time for {self.REPORT_LAST_SENT_DATETIME_CACHE_KEY} is "
+                f"{last_sent}, {td_last_sent.total_seconds()} seconds ago. "
+                "This is less than the minimum time between execution "
+                f"({self.min_time_between_loop_seconds} sec), so skipping this report..."
+            )
+            return
 
-    return inner
+        await send_weekly_contribution_reports(self.bot)
```

Comment:
Would we want to check the cadence before running weekly. What if the cadence is daily?

Reply:
can you explain differently?